lib/libc/sys/ioctl.c

```diff
@@ -88,9 +92,23 @@ _ioctl(int fd, unsigned long com, ...)
 		} else {
 			data = va_arg(ap, void *);
 		}
-		va_end(ap);
 	} else {
-		data = NULL;
+#if defined(__CHERI_PURE_CAPABILITY__) && __mips__
+		/*
+		 * For some ioctls, the size is zero, but the handler still
+		 * expect an address. This is done to skip the copyin in
+		 * user_ioctl, and handle it manually in the ioctl handler.
+		 */
+		if (cheri_getlen((void*)ap) >= sizeof(void *))
```

Comment:
```C
if (ap != NULL && cheri_getlen((void*)ap) >= sizeof(void *))
```

But in fact, I think the whole `#if`/`#else`/`#endif` can just be:

```
if ((void *)ap != NULL)
    data = (void *)va_arg(ap, syscallarg_t);
else
    data = NULL;
```

We only expect pointers or no arguments for size-zero ioctls, so with the NULL check we don't need to query the length. This will also work on CHERI RISC-V.

Reply:
Does `va_start` return a `void *` on all FreeBSD ABIs?

Reply:
No, it's a struct with a series of pointers on some of them... so we probably want:

```
#ifdef __CHERI_PURE_CAPABILITY__
if ((void *)ap == NULL)
    data = NULL;
else
#endif
    data = (void *)va_arg(ap, syscallarg_t);
```

and then worry about Morello later (AArch64 uses a struct, I assume so will Morello).

Reply:
Maybe we can fix this nonsense by the time we're ready to merge Morello support...